pkg/collector/corechecks/cluster/orchestrator/collectors/k8s/ingress.go

```diff
@@ -0,0 +1,136 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build kubeapiserver && orchestrator
+// +build kubeapiserver,orchestrator
+
+package k8s
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/cluster/orchestrator/collectors"
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/cluster/orchestrator/processors"
+	k8sProcessors "github.com/DataDog/datadog-agent/pkg/collector/corechecks/cluster/orchestrator/processors/k8s"
+	"github.com/DataDog/datadog-agent/pkg/orchestrator"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/datadog-agent/pkg/util/retry"
+
+	netv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	netv1Informers "k8s.io/client-go/informers/networking/v1"
+	netv1Listers "k8s.io/client-go/listers/networking/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+// IngressCollector is a collector for Kubernetes Ingresss.
+type IngressCollector struct {
+	informer    netv1Informers.IngressInformer
+	lister      netv1Listers.IngressLister
+	metadata    *collectors.CollectorMetadata
+	processor   *processors.Processor
+	retryLister func(ctx context.Context, opts metav1.ListOptions) (*netv1.IngressList, error)
+}
+
+// NewIngressCollector creates a new collector for the Kubernetes Ingress
+// resource.
+func NewIngressCollector() *IngressCollector {
+	return &IngressCollector{
+		metadata: &collectors.CollectorMetadata{
+			IsStable: false,
+			Name:     "ingresses",
+			NodeType: orchestrator.K8sIngress,
+		},
+		processor: processors.NewProcessor(new(k8sProcessors.IngressHandlers)),
+	}
+}
+
+// Informer returns the shared informer.
+func (c *IngressCollector) Informer() cache.SharedInformer {
+	return c.informer.Informer()
+}
+
+// Init is used to initialize the collector.
+func (c *IngressCollector) Init(rcfg *collectors.CollectorRunConfig) {
+	c.informer = rcfg.APIClient.InformerFactory.Networking().V1().Ingresses()
+	c.lister = c.informer.Lister()
+	c.retryLister = rcfg.APIClient.Cl.NetworkingV1().Ingresses("").List
+}
+
+// IsAvailable returns whether the collector is available.
+// Returns false if the networking.k8s.io/v1 API version is not available (kubernetes < 1.19).
+func (c *IngressCollector) IsAvailable() bool {
+	var retrier retry.Retrier
+	if err := retrier.SetupRetrier(&retry.Config{
+		Name:          "NetworkV1Discovery",
+		AttemptMethod: c.list,
```

Comment:
Yes ideally the discovery client can be used for this. We had exactly the same need for the admission controller, we relied on the discovery client but then we had to switch to this approach. See full details in this [PR](https://github.com/DataDog/datadog-agent/pull/10171).
TL;DR The discovery clients expects answers from all registered APIs including `external.metrics.k8s.io/v1beta1` which is the role of the DCA to answer it when the external metrics provider is enabled, at startup the DCA is not ready (expected) so every discovery client request would fail.
The current approach is not perfect indeed but it's robust, the expensive list operation would only be executed 3 times worst case.

Reply:
😮 

I'm not 100% sure if it means that:
- the DCA can block all responses to `/apis` from being returned for 30s
or
- the API for `external.metrics.k8s.io/v1beta1` doesn't show up in the list for 30s

Reply:
It's the former, `30s` is the best case (minimum time required to be come ready). During that time (DCA not ready) the discovery requests fail with this error `unable to retrieve the complete list of server APIs: external.metrics.k8s.io/v1beta1: the server is currently unable to handle the request`. It's a shame that the discovery client doesn't return a partial list of the available APIs in this case, it's all or nothing.

Reply:
Thanks for clarifying, then it's rather disappointing 😕

Not in the scope of that PR but maybe there's something we could do better in the external metrics provider to stop this from happening? I suppose we'll hit this issue again if we try to collect CRDs later.

Let's start with your approach for now 👍

Reply:
> but maybe there's something we could do better in the external metrics provider to stop this from happening?

It's a tough one, not ready pods don't serve service requests be design in k8s, and I'm not sure we want to remove the readiness prove for the cluster agent, so no obvious solution unfortunately. I'll add something to the backlog though to investigate deeper and try to find other alternatives.